src/coap.c

```diff
@@ -0,0 +1,313 @@
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ * |Ver| T | TKL | Code | Message ID | Token (if any, TKL bytes) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ * | Options (if any) ...            |1 1 1 1 1 1 1 1| Payload (if any) ...
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
+ */
+
+/* TODO(alashkin): remove this #include after dev completion */
+#include <stdlib.h>
+
+#include "coap.h"
+
+/* Options memory management functions */
+void ns_coap_free_options(struct ns_coap_message *cm) {
+  while (cm->options != NULL) {
+    struct ns_coap_option *next = cm->options->next;
+    free(cm->options);
+    cm->options = next;
+  }
+}
+
+void ns_coap_add_option(struct ns_coap_message *cm,
+                        uint16_t number, char* value, size_t len) {
+  struct ns_coap_option *option =
+      (struct ns_coap_option *)NS_CALLOC(1, sizeof(*option));
+
+  option->number = number;
+  option->value.p = value;
+  option->value.len = len;
+
+  if (cm->options == NULL) {
+    cm->options = cm->_options_tail = option;
+  } else {
+    cm->_options_tail = cm->_options_tail->next = option;
+  }
+}
+
+/* Fills CoAP header in cm. */
+static char *coap_parse_header(char* ptr, struct iobuf *io,
+                               struct ns_coap_message *cm) {
+  if (io->len < sizeof(uint32_t)) {
+    cm->flags |= COAP_NOT_ENOUGH_DATA;
+    return NULL;
+  }
+
+  /*
+   * Version (Ver):  2-bit unsigned integer.  Indicates the CoAP version
+   * number.  Implementations of this specification MUST set this field
+   * to 1 (01 binary).  Other values are reserved for future versions.
+   * Messages with unknown version numbers MUST be silently ignored.
+   */
+  if ((*ptr >> 6) != 1) {
+    cm->flags |= COAP_IGNORE;
+    return NULL;
+  }
+
+  /*
+   * Type (T):  2-bit unsigned integer.  Indicates if this message is of
+   * type Confirmable (0), Non-confirmable (1), Acknowledgement (2), or
+   * Reset (3).
+   */
+  cm->msg_type = (*ptr & 0x30) >> 4;
+  if (cm->msg_type > COAP_MSG_MAX) {
+    cm->flags |= COAP_FORMAT_ERROR;
+    return NULL;
+  }
+  cm->flags |= COAP_MSG_TYPE_FIELD;
+
+  /*
+   * Token Length (TKL):  4-bit unsigned integer.  Indicates the length of
+   * the variable-length Token field (0-8 bytes).  Lengths 9-15 are
+   * reserved, MUST NOT be sent, and MUST be processed as a message
+   * format error.
+   */
+  cm->token.len = *ptr & 0x0F;
+  if (cm->token.len > 8) {
+    cm->flags |= COAP_FORMAT_ERROR;
+    return NULL;
+  }
+
+  ptr++;
+
+  /*
+   * Code:  8-bit unsigned integer, split into a 3-bit class (most
+   * significant bits) and a 5-bit detail (least significant bits)
+   */
+  cm->code_class = *ptr >> 5;
+  cm->code_detail = *ptr & 0x1F;
+  cm->flags |= (COAP_CODE_CLASS_FIELD | COAP_CODE_DETAIL_FIELD);
+
+  ptr++;
+
+  /*
+   * Message ID:  16-bit unsigned integer in network byte order.
+   */
+  cm->msg_id = ntohs(*(uint16_t*)ptr);
```

Comment:
Yes. But what would you like to see there?
I suppose, get/put_unaligned is not an options  by reason of portability.
`ntohs((uint8_t)*(ptr+1) << 8 | (uint8_t)*ptr)`?